ARROW-7921: [Go] Add Reset method to various components and clean up comments.

richardartoul opened the pull request:
The reset method allow the data structures to be re-used so they don't have to be allocated over and over again.

wesm commented:
Can you please open a JIRA issue for the benefit of our changelog?

sbinet commented on go/arrow/array/string.go:
```diff
@@ -30,7 +30,7 @@ const (
 )
 
-// A type which represents an immutable sequence of variable-length UTF-8 strings.
+// String is a type which represents an immutable sequence of variable-length UTF-8 strings.
```
```suggestion
// String represents an immutable sequence of variable-length UTF-8 strings.
```

sbinet commented on go/arrow/array/data.go:
```diff
@@ -24,7 +24,7 @@ import (
 )
 
-// A type which represents the memory and metadata for an Arrow array.
+// Data is a type which represents the memory and metadata for an Arrow array.
```
```suggestion
// Data represents the memory and metadata of an Arrow array.
```

sbinet commented on go/arrow/memory/buffer.go:
```diff
@@ -60,15 +61,28 @@ func (b *Buffer) Release() {
 }
 
+// Reset resets the buffer for reuse.
+func (b *Buffer) Reset(buf []byte) {
```
it's a part of the codebase I am not completely fluent with.
but, shouldn't we also somewhat test whether that `Buffer` is `Mutable()` before modifying it?

richardartoul replied:
As far as I can tell `mutable` just means `there is a reference to an allocator here so things can be resized`, I don't think it has much bearing on calling `Reset()`.

The `Reset()` is used to reuse data structures so you're also kind of already saying: "I know what I'm doing"

sbinet replied:
ok

sbinet commented:
also, it would be nice to have a few tests for this new API.

richardartoul commented:
@sbinet I think I addressed all your feedback. I updated the method strings and added some tests. Not really sure what to do about the mutability thing (I left my thoughts on that) but I don't think it impacts correctness at all and the new methods are mostly optimizations to prevent extra allocations for people who want to eak out some extra perf in the hot path.

richardartoul commented:
@wesm Opened a ticket!
